core/Enclave/enclave_hsm.cpp

```diff
@@ -107,21 +94,59 @@ void printf(const char *fmt, ...)
     ocall_print_string(buf);
 }
 
-static uint32_t sgx_calc_gcm_data_size(const uint32_t aad_size, const uint32_t plaintext_size)
+static uint32_t ehsm_calc_gcm_data_size(const uint32_t aad_size, const uint32_t plaintext_size)
 {
-    if (aad_size > UINT32_MAX - sizeof(aes_gcm_key_data_t))
+    if (aad_size > UINT32_MAX - sizeof(sgx_aes_gcm_data_ex_t))
         return UINT32_MAX;
 
-    if (plaintext_size > UINT32_MAX - sizeof(aes_gcm_key_data_t))
+    if (plaintext_size > UINT32_MAX - sizeof(sgx_aes_gcm_data_ex_t))
         return UINT32_MAX;
 
     if (aad_size > UINT32_MAX - plaintext_size)
         return UINT32_MAX;
 
-    if (sizeof(aes_gcm_key_data_t) > UINT32_MAX - plaintext_size - aad_size)
+    if (sizeof(sgx_aes_gcm_data_ex_t) > UINT32_MAX - plaintext_size - aad_size)
+        return UINT32_MAX;
+
+    return (aad_size + plaintext_size + sizeof(sgx_aes_gcm_data_ex_t));
+}
+
+static uint32_t sgx_get_gcm_ciphertext_size(const sgx_aes_gcm_data_ex_t *gcm_data)
+{
+    if (NULL == gcm_data)
         return UINT32_MAX;
 
-    return (aad_size + plaintext_size + sizeof(aes_gcm_key_data_t));
+    return gcm_data->ciphertext_size;
+}
+
+static uint32_t ehsm_get_rsa_key_pem_size(const uint32_t keyspec)
+{
+    switch (keyspec)
+    {
+        case EH_RSA_2048:
+            return RSA_2048_PUBLIC_KEY_PEM_SIZE + RSA_2048_PRIVATE_KEY_PEM_SIZE;
+        case EH_RSA_3072:
+            return RSA_3072_PUBLIC_KEY_PEM_SIZE + RSA_3072_PRIVATE_KEY_PEM_SIZE;
+        case EH_RSA_4096:
+            return RSA_4096_PUBLIC_KEY_PEM_SIZE + RSA_4096_PRIVATE_KEY_PEM_SIZE;
+        default:
+            return UINT32_MAX;
```

Comment:
if the default error is UINT32_MAX, every places should check the return value.
@322, @704

Reply:
Is 0 a valid value?  If not, you can use 0 as the default value to indicate an error.

Reply:
This function will use in ehsm_calc_gcm_data_size(add_size, plaintext_size) to get keybloblen
And UINT32_MAX in it will return error